### ISSUE TITLE
About the required whitespace before a colon in type restrictions

### ISSUE THREAD

**Issue by Dark565** [crystal-lang/crystal#10698]
Gentlemen, I know that the thing I'm talking about is a huge detail, but in my opinion greatly changing the final feelings of the language.

What is the practical reason of requiring a space before a colon in type restrictions?

If it is because the structure of the parser - it was needed "as a hack" to keep parser working at a time and remained to this day, then it is somehow reasonable.
If it is because 'wanting to make the language more readable" then unfortunately it's not very so.

I can list a lot of popular and loved languages that don't have such a requirement and yet nobody cares.
Take for example:
Rust:
```rust
fn foo(x: u32, y: f32);
```
Nim:
```nim
proc foo(x: Natural, y: float)
```
TypeScript
```typescript
function foo(x: number, y: number)
```
Zig
```zig
fn foo(x: u32, y: f32) void
```
And many more.

`x : Type` syntax doesn't make it more readable.
It just increases the number of characters that a programmer has to enter and makes code more scattered.
Lack of space before ':' looks more natural if we took writing of our tongue languages like English.
We write `some list:` instead `some list :`.
Additionally code editors may make it even more sugar by highlighting name of a variable and a colon in different colors.
Dense ':' clearly shows that type assignment is something important as it is for the language that is natively compiled.

What are your opinions about that?

**Comment by straight-shoota** [comment]
I agree that whitespace might not be necessary for readabilty. However, I don't buy the arguments why no whitespace would supposedly be superior.
More characters to type is hardly an issue. We're talking about a single space.
Familiarity with the "lack of space" is subjective. If you personally don't like it, maybe you can configure your editor to hide the space? =)

Besides, there is a semantic difference: `x: Foo` means a named argument `x` with value `Foo`, and `x : Foo` declares the variable `x` is of type `Foo`. This distinction is important. For example: `property x: Foo` would pass a named argument to the `property` macro. `property x : Foo` passes a type declaration (which is the expected usage of this specific macro).

**Comment by asterite** [comment]
The answer is very simple:

`x : Int32` is a type declaration. It declares `x` to be of type `Int32`

`x: Int32` is a named argument named "x" with the value `Int32`

Consider this code:

```crystal
class Person
  # This calls the property macro and passes the type declaration `name : String`
  property name : String

  # This calls the property macro and passes a named argument `name` with value `String`
  # This one doesn't work. You get an error: no argument named 'name'
  property name: String
end
```

**Comment by asterite** [comment]
Oh, sorry, after I wrote all of that I see that @straight-shoota already answered the same thing. Sorry 🙏

**Comment by Dark565** [comment]
Thanks for the answers.
I forgot to take the complications with macros (that can be invoked like `my_macro x: name : Int32`) into respect.
We could allow no space between `:` in the type restrictions in method parameters, but later it would make confusion whether doing `my_macro x: Int32` actually passes `x of type Int32` or `parameter x of a macro with value Int32` to the macro.
It would be needed to specify pointless syntax exceptions like `in macros, space before ':' is needed to specify a type`.
Now that you reminded about that, current syntax indeed has sense.

However, by the way, I see, Crystal doesn't allow to specify the default values of arguments in methods with the named arguments syntax.
Something that Ruby has supported since 2.0. Precisely I'm talking about:
```ruby
def foo(x: 32, y: 64)
```
What is the reason of this lack?
It would probably conflict with nothing and more resemble Ruby.

> BTW, @Dark565 we're not all gentlemen here ;-)

@beta-ziliani Yeah, sorry for that 😅

**Comment by Blacksmoke16** [comment]
@Dark565 https://crystal-lang.org/reference/tutorials/basics/60_methods.html#default-arguments.  `var = "default"`.

**Comment by Dark565** [comment]
@Blacksmoke16 No need to link the reference. I obviously know about '='.
I'm just asking about the specific ':' behavior, which could be made optional as in Ruby where we can do it by either `=` and `:`.

**Comment by Blacksmoke16** [comment]
@Dark565 I imagine it would conflict with the case where the argument is typed. I.e. `def(x : Int32)`.  `def(x: 32 : Int32` is getting a bit out of hand.  I could see it working in the untyped context, but at that point there's not really a benefit using `:` versus `=` given the latter keeps things standardized.

**Comment by asterite** [comment]
Also:

```crystal
def foo(x: Int32)
```

Is that a named argument `x` with a default **value** of `Int32`, or a regular argument `x` with a type `Int32`?

**Comment by HertzDevil** [comment]
The `:` is not optional in Ruby; it delimits named parameters from positional parameters, and does _not_ refer to a default argument on its own. So the following in Crystal:

```crystal
def foo(a, b = 1, *, c, d = 1); end
```

is really equivalent to:

```ruby
def foo(a, b = 1, c:, d: 1); end
```

And honestly I think Crystal does it better. Ruby does not allow `d: = 1`.

***

By the way, it is entirely possible to write your own `property` macro that takes named arguments:

```crystal
class Object
  macro my_property(**opts)
    {% for name, value in opts %}
      @{{ name.id }} : {{ value }}
      
      def {{ name.id }}
        @{{ name.id }}
      end
      
      def {{ name.id }}=(@{{ name.id }})
      end
    {% end %}
  end
end

class A
  my_property x: Int32, y: Int32
  my_property z: Bool
  
  def initialize
    @x = 1
    @y = 2
    @z = true
  end
end
```

As you might already notice, this prevents the macro from taking an initializer for the property. That's why `property` takes a type declaration instead of a named argument.

**Comment by Dark565** [comment]
@asterite The way I saw it was 
```crystal
def foo(x: Int32)
```
means "x with the default value Int32", while
```crystal
def foo(x : Int32: 32)
```
"x of type Int32 with the default value 32".
It would be not permitted by the compiler in that case to write something like
```crystal
def foo(x: 32 : Int32)
```
similarly like it's not possible to write
```crystal
def foo(x = 32 : Int32)
```
---
@HertzDevil Good that you pointed that.
My error that I treated ':' in the argument field like the exact synonym of '=' in this case.
Now as you corrected me, my suggestion is no longer reasonable.
`*` character takes away the need of `:`.
If one was to persist, it still could be used as an alternative for '=' as I suggested above, but it doesn't make any sense taking into account that it's not based on Ruby's syntax and defining one more behavior backed by nothing will definitely lead to confusion.
One could also reimplement
```crystal
def foo(x, y, *, z, w)
```
as
```crystal
def foo(x, y, z:, w:)
```
like in Ruby, but 

- it's slightly too late (it's 1.0 now)
- ones can argue whether Ruby's or Crystal's syntax of named parameters is better, but it's already good and adopted; it isn't sensible to change it except of desire of keeping extremely strong compatibility with the Ruby code, which isn't the case.
---
Now, as everything is clean, I think I can close this issue.
Thanks for the discussion :)

**Comment by HertzDevil** [comment]
It turns out block parameters don't require a space before the colon:

```crystal
def foo(&block: Int32 -> Char) # okay
end

def bar(&: Int32 -> Char) # okay
end
```

It can be seen in use [here](https://github.com/crystal-lang/crystal/blob/f90d914f90191c3035681e9a585983abf1af7a59/spec/compiler/semantic/block_spec.cr#L153). The parser and the formatter could handle this, but for consistency I think we should make this space mandatory too.

**Comment by genki** [comment]
We know the grammar of Crystal respects to Ruby's.
But there is a deadly difference between them, that is existence of the type specification.
Ruby's grammar is dense enough without the type specification so that Crystal is too tight if keep sticking to mimic Ruby.
Can't there be depart partially from Ruby?
I think it is better for Crystal to use '=' for named argument instead of ':' as like as Python.
As far as I know, Python's style is better to coexist with type specification.

**Comment by straight-shoota** [comment]
@genki I don't think there's much reason to do that. I think the differentiation between named argument and type restriction is quite clear (even though it's just a single whitespace in front of the colon).

Crystal clearly doesn't have to stick to Ruby. But at this point, it has to stick to its own history with a colon for named argument assignments.

**Comment by genki** [comment]
@straight-shoota 
It is true if we use only Crystal for programming, but I am not,
so I have to worry always about if I forget to put a white space in front of ':'
It's a bit confusing.
